lib/utils/errors.ts

```diff
@@ -0,0 +1,41 @@
+/**
+ * Contains Error types,
+ * used to clarify the intent of a throw
+ * and useful inside unit tests.
+ */
+
+export const enum ErrorTypes {
+  TransientError = "TransientError",
+  PermanentError = "PermanentError"
+}
+
+interface ICustomError<T extends ErrorTypes> {
+  readonly kind: T;
+  readonly message: string;
+  readonly cause?: Error;
+}
+
+const CustomError = <T extends ErrorTypes>(
```

Comment:
not sure why you need the type parameter `T`, I don't see any code that needs to be generic in `T` - it looks like the following would be enough:

``` ts
interface ICustomError {
   readonly kind: ErrorTypes;
   readonly message: string;
   readonly cause?: Error;
 }
```

Reply:
ok

Reply:
maybe needed in the future === not needed 😸

Reply:
I would rename `CustomError` to `RuntimeError`, seems more effective in communicating the intent